lib/configSchema.js

```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const functionNamePattern = '^[a-zA-Z0-9-_]+$';
+
+const schema = {
+  type: 'object',
+  properties: {
+    service: {
+      type: 'object',
+      properties: {
+        name: { pattern: '^[a-zA-Z][0-9a-zA-Z-]+$' },
+        awsKmsKeyArn: { pattern: '^arn:(aws[a-zA-Z-]*)?:kms:[a-z0-9-]+-\\d+:\\d{12}:[^\\s]+$' },
+      },
+      additionalProperties: false,
+    },
+
+    custom: {
+      type: 'object',
+      properties: {},
+      required: [],
+      additionalProperties: true, // user is free to add properties that he or she wants
+    },
+    plugins: { type: 'array', items: { type: 'string' } },
+
+    // TODO: Describe resources schema
+    resources: { type: 'object' },
+
+    functions: {
+      type: 'object',
+      patternProperties: {
+        [functionNamePattern]: {
+          type: 'object',
+          properties: {
+            name: { type: 'string' }, // name property is added by service class
+            events: {
+              type: 'array',
+              items: {}, // event plugins push their schemas to this array
+            },
+          },
+
+          /*
+           * Schema for function properties is extended by a provider plugin.
+           * For example, in the context of AWS provider props like
+           * provisionedConcurrency, memorySize, reservedConcurrency, etc.
+           * should be extended by awsProvider plugin.
+           */
+          additionalProperties: false,
+        },
+      },
+      additionalProperties: false,
+    },
+
+    /*
+     * Provider specific properties are extended in respected provider plugins.
+     */
+    provider: {
+      type: 'object',
+      properties: {},
+      required: ['name'],
+      additionalProperties: false,
+    },
+
+    package: {
+      type: 'object',
+      properties: {
+        individually: { type: 'boolean' },
+        path: { type: 'string' },
+        artifact: { type: 'string' },
+        exclude: { type: 'array', items: { type: 'string' } },
+        include: { type: 'array', items: { type: 'string' } },
+        excludeDevDependencies: { type: 'boolean' },
+      },
+      additionalProperties: false,
+    },
+
+    // TODO: Describe layers schema
+    layers: { type: 'object' },
+
+    configValidationMode: { enum: ['error', 'warn'] }, // in case of 'error' an exception is thrown
```

Comment:
Other thing that came to my mind.

I think we also need to introduce `'off'` option. I believe there can be case where:
- Users will rely on some unmaintaned plugins that will tweak the config, which will produce error, and yet there'll be no one to define schemas.
- We may have some bugs and report false positives (while they should be quickly addressed, users may have reason to stick older versions of a Framework)

Users may want to turn off validation to avoid noise while they work. When `'off'` is set, I would do not follow with any validation (just abort right at the beginning of `service.validate`)

Reply:
I think this will be useful and also easy to implement. I'll add it now

Reply:
Implemented in a bit different way, see https://github.com/serverless/serverless/pull/7335/commits/27efc60daadd7326e18fcf26bc82dbef3535b20d. This also works

Reply:
Looks great! It'll be nice to also mention this option here: https://github.com/serverless/serverless/blob/57e4212671ea3027fab9482e6006933e4c5b6c55/docs/providers/aws/guide/serverless.yml.md

Reply:
Sure will do that